throw bad request exception when validation fails

shharma-vipin opened the pull request:
This  PR is to handle issue https://github.com/hwslabs/iam/issues/51

Fix include changes in following files :
	* **RequestModeValidatorKt
	* ConstantsKt
	* UserApiTestKt (existing test failed after adding password validation + added negative scenario test case)**

Added validation support for following entities :
 > **Password** :
 	  regex_expression -> `^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@_#])[A-Za-z\d@_#]{8,}$`
	  description -> Password should contain at least one uppercase letter, one lowercase letter, one number and one special character[@ _ #]
	  
 > **Email** :
 	  regex_expression -> `^\S+@\S+\.\S+$`
	  description -> Email should contain `.`, `@`
	  
 > **Phone Number** :
 	  regex_expression -> `^[+]*[(]{0,1}[0-9]{1,4}[)]{0,1}[-\s\./0-9]*$`
	  description -> Only characters +, -, 0..9 are supported.

shharma-vipin commented:
I missed running test cases, which is why action is failing. Allow me sometime to fix that.

hari-hypto commented on src/main/kotlin/com/hypto/iam/server/extensions/Validators.kt:
```diff
@@ -67,7 +68,7 @@ val nameCheck = Validation<String> {
 fun <T> Validation<T>.validateAndThrowOnFailure(value: T): T {
     val result = validate(value)
     if (result is Invalid<T>) {
-        throw IllegalArgumentException(result.errors.toString())
```
IllegalArgumentException is already converted as 400 Bad request.

https://github.com/hwslabs/iam/blob/main/src/main/kotlin/com/hypto/iam/server/ExceptionHandler.kt#L70

Fix for this issue #51 would be to add input validation in "CreateOrganization" & "CreateUser" APIs.
1. Password length (minimum 8 chars length) & Password format (1 uppercase, 1 lowercase, one special char @_#)
2. Email format validation (<username>@<domain>)
3. Phone number validation (+?<8 or more digits with support of special "-" hyphen character>)

shharma-vipin replied:
@hari-hypto Thank you for comments.
I tried running tests in local environment & found out that, on validation failure response status returned as 400 (which is expected in this case), however issue #51  mentioned that it is returning "unknown server error"
Please let me know incase I am missing something here.

hari-hypto replied:
@shharma-vipin 
The issue here is our input validation is too lenient for password & phone number. So there were cases when the input validation was successful but exception was returned by our user management service layer (AWS Cognito) for password & phone number.

Since any failures from service layer is returned as "Internal Server Error", it was not conveyed to clients properly. So the solution is to have a more stricter validation at api layer itself so that service layer will not reject this input.

Overall the final PR version looks good to me. One minor comment is to add this validation check even for the "AdminUser" when creating organization.

hari-hypto commented:
> I missed running test cases, which is why action is failing. Allow me sometime to fix that.

Sure. Thanks for raising PR. I added some comments. Please check and reply back if you have any questions. We are also available in [slack](https://join.slack.com/t/hypto-community/shared_invite/zt-160bce60l-5cIm2qJFaJP8rs3gKHRLLg/) and we can discuss there as well.

shharma-vipin commented:
@hari-hypto 
Github Action failing with following message
```
Push to branch badges
remote: Permission to hwslabs/iam.git denied to github-actions[bot].
fatal: unable to access 'https://github.com/hwslabs/iam.git/': The requested URL returned error: 403
Error: Invalid exit code: 128
    at ChildProcess.<anonymous> (/home/runner/work/_actions/ad-m/github-push-action/master/start.js:29:21)
    at ChildProcess.emit (events.js:314:20)
    at maybeClose (internal/child_process.js:1022:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:28[7](https://github.com/hwslabs/iam/runs/5752370899?check_suite_focus=true#step:9:7):5) {
  code: 12[8](https://github.com/hwslabs/iam/runs/5752370899?check_suite_focus=true#step:9:8)
}
Error: Invalid exit code: 128
```

hari-hypto commented:
> Push Commit Badge Github Action failing with following message

Seems to be a permission issue. Looking into it. We will get back.

hari-hypto reviewed (commented):
The change looks good to me. One minor comment is to add this validation check even for the "AdminUser" password and phoneNumber when creating organization.

shharma-vipin commented:
> The change looks good to me. One minor comment is to add this validation check even for the "AdminUser" password and phoneNumber when creating organization.

sure @hari-hypto.
I've added validations for admin details like (password, phone, username, email). And fixed failing test cases after adding validations

shharma-vipin commented:
@hari-hypto There was code smell with wildcard import causing github action to fail.
Request you to trigger build action again.

hari-hypto commented:
> @hari-hypto There was code smell with wildcard import causing github action to fail. Request you to trigger build action again.

Thanks for fixing it. Pushed the PR changes to main branch.
